File: examples/Simple/client/rb/Dynamic/SimpleClient.rb
#!/usr/bin/env ruby

$LOAD_PATH.push('../gen-rb')

require 'thrift'
require 'simple_service'

def check(data_out, expected, size)
  status = 0
  (0..(size - 1)).each do |i|
    if data_out[i] != expected[i]
      puts "#{data_out[i]} != #{expected[i]}"
      status += 1
    end
  end
  status
end

def simple_cpu(size, data_in, data_out)
  (0..(size - 1)).each do |i|
    data_out[i] = (data_in[i] * data_in[i] + data_in[i])
  end
end

begin
  include Com::Maxeler::Simple

  start_time = Time.now
  start_dfe_time = start_time

  port = 9090

  # Make socket
  socket = Thrift::Socket.new('localhost', port)

  # Buffering is critical. Raw sockets are very slow
  transport = Thrift::BufferedTransport.new(socket)

  # Wrap in a protocol
  protocol = Thrift::BinaryProtocol.new(transport)

  # Create a client to use the protocol encoder
  client = SimpleService::Client.new(protocol)

  current_time = (Time.now - start_time).round(5)
  puts "Creating a client:\t\t\t\t#{current_time}s"

  # Connect!
  start_time = Time.now
  transport.open
  current_time = (Time.now - start_time).round(5)
  puts "Opening connection:\t\t\t\t#{current_time}s"

  start_time = Time.now
  size = 384
  size_bytes = 4 * size

  # Generate input
  data_in = Array.new(size)
  (0..(size - 1)).each do |i|
    data_in[i] = i + 1
  end
  current_time = (Time.now - start_time).round(5)
  puts "Generating input data:\t\t\t\t#{current_time}s"

  # Initialize maxfile
  start_time = Time.now
  max_file = client.Simple_init
  current_time = (Time.now - start_time).round(5)
  puts "Initializing maxfile:\t\t\t\t#{current_time}s"

  # Load DFE
  start_time = Time.now
  max_engine = client.max_load(max_file, '*')
  current_time = (Time.now - start_time).round(5)
  puts "Loading DFE:\t\t\t\t\t#{current_time}s"

  # Allocate and send input streams to server
  start_time = Time.now
  address_data_in = client.malloc_float(size)
  client.send_data_float(address_data_in, data_in)
  current_time = (Time.now - start_time).round(5)
  puts "Sending input data:\t\t\t\t#{current_time}s"

  # Allocate memory for output stream on server
  start_time = Time.now
  address_data_out = client.malloc_float(size)
  current_time = (Time.now - start_time).round(5)
  puts "Allocating memory for output stream on server:\t#{current_time}s"

  # Action default
  start_time = Time.now
  actions = client.max_actions_init(max_file, 'default')
  client.max_set_param_uint64t(actions, 'N', size)
  client.max_queue_input(actions, 'x', address_data_in, size_bytes)
  client.max_queue_output(actions, 'y', address_data_out, size_bytes)
  client.max_run(max_engine, actions)
  current_time = (Time.now - start_time).round(5)
  puts "Simple time:\t\t\t\t\t#{current_time}s"

  # Unload DFE
  start_time = Time.now
  client.max_unload(max_engine)
  current_time = (Time.now - start_time).round(5)
  puts "Unloading DFE:\t\t\t\t\t#{current_time}s"

  # Get output stream from server
  start_time = Time.now
  data_out = client.receive_data_float(address_data_out, size)
  current_time = (Time.now - start_time).round(5)
  puts "Getting output stream:\t(size = #{size * 32} bit)\t#{current_time}s"

  # Free allocated memory for streams on server
  start_time = Time.now
  client.free(address_data_in)
  client.free(address_data_out)
  client.free(actions)
  current_time = (Time.now - start_time).round(5)
  puts "Freeing allocated memory for streams on server:\t#{current_time}s"

  # Free allocated maxfile data
  start_time = Time.now
  client.Simple_free
  current_time = (Time.now - start_time).round(5)
  puts "Freeing allocated maxfile data:\t\t\t#{current_time}s"

  # Close!
  start_time = Time.now
  transport.close
  current_time = (Time.now - start_time).round(5)
  puts "Closing connection:\t\t\t\t#{current_time}s"

  current_time = (Time.now - start_dfe_time).round(5)
  puts "DFE simple total time:\t\t\t\t#{current_time}s"

  # CPU Output
  start_time = Time.now
  expected = Array.new(size)
  simple_cpu(size, data_in, expected)
  current_time = (Time.now - start_time).round(5)
  puts "CPU simple total time:\t\t\t\t#{current_time}s"

  # Checking results
  start_time = Time.now
  status = check(data_out, expected, size)
  current_time = (Time.now - start_time).round(5)
  puts "Checking results:\t\t\t\t#{current_time}s"

  if status == 0
    puts 'Test passed!'
  else
    puts "Test failed #{status} times!"
    Kernel.exit(-1)
  end
rescue Thrift::Exception => thrift_exception
  puts 'Thrift::Exception: ', thrift_exception.message, "\n"
  Kernel.exit(-1)
end

File: examples/Simple/client/cpp/Dynamic/SimpleClient.cpp
#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>
#include <stdlib.h>

#include "../gen-cpp/SimpleService.h"

using namespace std;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;

using namespace ::com::maxeler::Simple;

int check(std::vector<double> dataOutDFE, std::vector<double> dataOutCPU, int size) {
	int status = 0;

	for(int i=0; i < size; i++) {
		if(dataOutDFE[i] != dataOutCPU[i]) {
			fprintf(stderr, "Output data @ %d = %1.8g (expected %1.8g)\n",
				i, dataOutDFE[i], dataOutCPU[i]);
			status = 1;
		}
	}

	return status;
}

std::vector<double> SimpleCPU(int size, std::vector<double> dataIn) {
	std::vector<double> dataOut;
	dataOut.resize(size);

	for (int i=0 ; i<size ; i++) {
		dataOut[i] = dataIn[i]*dataIn[i] + dataIn[i];
	}

	return dataOut;
}

std::vector<double> SimpleDFE(int size, std::vector<double> dataIn) {
	// Make socket
	boost::shared_ptr<TTransport> socket(new TSocket("localhost", 9090));

	// Buffering is critical. Raw sockets are very slow
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));

	// Wrap in a protocol
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));

	// Create a client to use the protocol encoder
	SimpleServiceClient client(protocol);

	int sizeBytes = size * sizeof(float);

	std::vector<double> dataOut;
	dataOut.resize(size);

	try {
		// Connect!
		transport->open();

		// Initialize maxfile
		remote_ptr maxfile = client.Simple_init();

		// Load DFE
		remote_ptr engine = client.max_load(maxfile, "*");

		// Allocate and send input streams to server
		remote_ptr address_dataIn = client.malloc_float(size);
		client.send_data_float(address_dataIn, dataIn);

		// Allocate memory for output stream on server
		remote_ptr address_dataOut = client.malloc_float(size);

		cout << "Running DFE.\n";

		remote_ptr actions = client.max_actions_init(maxfile, "default");
		client.max_set_param_uint64t(actions, "N", size);
		client.max_queue_input(actions, "x", address_dataIn, sizeBytes);
		client.max_queue_output(actions, "y", address_dataOut, sizeBytes);

		client.max_run(engine, actions);

		// Unload DFE
		client.max_unload(engine);

		// Get output stream from server
	        client.receive_data_float(dataOut, address_dataOut, size);

		// Free allocated memory for streams on server
		client.free(address_dataIn);
        	client.free(address_dataOut);
	
		// Free allocated maxfile data
		client.Simple_free();

		// Close!
		transport->close();
    
	} catch (TException& tx) {
		cout << "ERROR: " << tx.what() << endl;
	}

	return dataOut;
}

int main() {
	// Input
	const int size = 1024;
	std::vector<double> dataIn;

	for(int i = 0; i < size; i++) {
		dataIn.push_back(i + 1);
	}

	// CPU Output
	std::vector<double> dataOutCPU = SimpleCPU(size, dataIn);

	// DFE Output
        std::vector<double> dataOutDFE = SimpleDFE(size, dataIn);

	// Checking results
	if (check(dataOutDFE, dataOutCPU, size)) {
		cout << "Test failed." << endl;
	} else {
		cout << "Test passed!" << endl;
	}

	return 0;
}

